File: engines/clocks/binary_clock.py
from PIL import ImageDraw

class BinaryClock:
    def __init__(self, width, height):
        self.w = width
        self.h = height
        
    def tick(self, img, time_str, font, color1, color2):
        draw = ImageDraw.Draw(img)
        
        parts = time_str.split(':')
        if len(parts) >= 3:
            h, m, s = parts[0], parts[1], parts[2]
        else:
            h, m, s = "00", "00", "00"
            
        # Ensure 2 digits
        h = h.zfill(2)
        m = m.zfill(2)
        s = s.zfill(2)
        
        digits = [int(h[0]), int(h[1]), int(m[0]), int(m[1]), int(s[0]), int(s[1])]
        max_bits = [2, 4, 3, 4, 3, 4] # Max bits needed for each column
        
        dot_radius = max(2, min(self.w // 20, self.h // 12))
        spacing_x = self.w // 8
        spacing_y = self.h // 6
        
        start_x = (self.w - (5 * spacing_x)) // 2
        start_y = self.h - (self.h // 6) # Bottom aligned
        
        for col, val in enumerate(digits):
            x = start_x + col * spacing_x
            if col in [2, 4]: # Add extra gap between H/M and M/S
                x += spacing_x // 2
                start_x += spacing_x // 2
                
            for bit in range(max_bits[col]):
                y = start_y - bit * spacing_y
                is_on = (val >> bit) & 1
                
                # Draw dot
                if is_on:
                    c = color1 if col < 2 else (color2 if col < 4 else (255,255,255))
                    draw.ellipse([x - dot_radius, y - dot_radius, x + dot_radius, y + dot_radius], fill=c)
                else:
                    c = (30, 30, 30) # Dimmed
                    draw.ellipse([x - dot_radius, y - dot_radius, x + dot_radius, y + dot_radius], outline=c, width=1)
                    
        return img

File: src/engines/clocks/binary_clock.rs
use crate::core::matrix::MatrixBackend;
use crate::engines::renderers::base_renderer::ArcadeFont;

pub struct BinaryClock;

impl BinaryClock {
    pub fn new() -> Self {
        Self
    }

    pub fn render(
        &self,
        matrix: &mut dyn MatrixBackend,
        hours: u32,
        minutes: u32,
        seconds: u32,
        _font: &ArcadeFont<'_>,
        _scale: u32,
    ) {
        let w = matrix.width() as i32;
        let h = matrix.height() as i32;

        // 6 columns: H1 H2 M1 M2 S1 S2
        let digits = [
            hours / 10,
            hours % 10,
            minutes / 10,
            minutes % 10,
            seconds / 10,
            seconds % 10,
        ];
        // Max representable bits per column (BCD: tens are 0-2, units are 0-9)
        let max_bits: [u32; 6] = [2, 4, 3, 4, 3, 4];

        // Colors per group: Hours = color1 (cyan), Minutes = color2 (magenta), Seconds = white
        let colors: [(u8, u8, u8); 3] = [(0, 220, 255), (255, 0, 180), (200, 200, 200)];
        let dim_color: (u8, u8, u8) = (25, 25, 25);

        let dot_r = ((w / 20).max(2)).min(h / 12) as i32;
        let spacing_x = w / 8;
        let spacing_y = h / 6;

        // Center the 6 columns (with gap after col 1 and col 3)
        // Extra gap between H/M and M/S pairs
        let mut start_x = (w - 5 * spacing_x) / 2;
        let start_y = h - h / 6;

        for (col, &val) in digits.iter().enumerate() {
            // Extra horizontal gap between pairs (H|M and M|S)
            if col == 2 || col == 4 {
                start_x += spacing_x / 2;
            }
            let x = start_x + col as i32 * spacing_x;
            let color_idx = col / 2; // 0=H, 1=M, 2=S
            let fg = colors[color_idx];

            for bit in 0..max_bits[col] {
                let y = start_y - bit as i32 * spacing_y;
                let is_on = (val >> bit) & 1 == 1;

                if is_on {
                    // Filled circle for "on" bit
                    for dy in -dot_r..=dot_r {
                        for dx in -dot_r..=dot_r {
                            if dx * dx + dy * dy <= dot_r * dot_r {
                                matrix.set_pixel(x + dx, y + dy, fg.0, fg.1, fg.2);
                            }
                        }
                    }
                } else {
                    // Dim outline circle for "off" bit
                    for dy in -dot_r..=dot_r {
                        for dx in -dot_r..=dot_r {
                            let dist_sq = dx * dx + dy * dy;
                            if dist_sq <= dot_r * dot_r && dist_sq >= (dot_r - 1) * (dot_r - 1) {
                                matrix.set_pixel(
                                    x + dx,
                                    y + dy,
                                    dim_color.0,
                                    dim_color.1,
                                    dim_color.2,
                                );
                            }
                        }
                    }
                }
            }
        }
    }
}
